src/datasource/globus/credentials_provider.ts

```diff
@@ -0,0 +1,239 @@
+  url.searchParams.set("code_challenge_method", "S256");
+  url.searchParams.set("state", state);
+  url.searchParams.set("scope", scope.join(" "));
+  return url.toString();
+}
+
+function getGlobusTokenURL({
+  clientId,
+  code,
+  code_verifier,
+}: {
+  code: string;
+  clientId: string;
+  code_verifier: string;
+}) {
+  const url = new URL("/v2/oauth2/token", GLOBUS_AUTH_HOST);
+  url.searchParams.set("grant_type", "authorization_code");
+  url.searchParams.set("client_id", clientId);
+  url.searchParams.set("redirect_uri", REDIRECT_URI);
+  url.searchParams.set("code_verifier", code_verifier);
+  url.searchParams.set("code", code);
+  return url.toString();
+}
+
+type GlobusLocalStorage = {
+  authorizations?: {
+    [resourceServer: string]: OAuth2Credentials;
+  };
+  /**
+   * Globus Connect Server domain mappings.
+   * Currently, there is no way to progrmatically determine the UUID of a GCS
+   * endpoint from their domain name, so a user will need to provide a UUID
+   * when attempting to access a file from a GCS endpoint.
+   */
+  domainMappings?: {
+    [domain: string]: string;
+  };
+};
+
+function getStorage() {
+  return JSON.parse(
+    localStorage.getItem("globus") || "{}",
+  ) as GlobusLocalStorage;
+}
+
+async function waitForAuth(
+  clientId: string,
+  globusConnectServerDomain: string,
+): Promise<OAuth2Credentials> {
+  const status = new StatusMessage(false, true);
+
+  const res: Promise<OAuth2Credentials> = new Promise((resolve) => {
+    const frag = document.createDocumentFragment();
+    const wrapper = document.createElement("div");
+    wrapper.style.display = "flex";
+    wrapper.style.flexDirection = "column";
+    wrapper.style.alignItems = "center";
+    frag.appendChild(wrapper);
+
+    const msg = document.createElement("div");
+    msg.textContent = "You must log in to Globus to access this resource.";
+    msg.style.marginBottom = ".5em";
+    wrapper.appendChild(msg);
+
+    const link = document.createElement("button");
+    link.textContent = "Log in to Globus";
+
+    link.addEventListener("click", async (event) => {
+      event.preventDefault();
+      /**
+       * We make a request to the Globus Connect Server domain **even though we _know_ we're
+       * unauthorized** to get the required consents for the resource.
+       */
+      const authorizationIntrospectionRequest = await fetch(
+        globusConnectServerDomain,
+        {
+          method: "GET",
+          headers: {
+            "X-Requested-With": "XMLHttpRequest",
+          },
+        },
+      );
+
+      const { authorization_parameters } =
+        await authorizationIntrospectionRequest.json();
+
+      const verifier = generateCodeVerifier();
+      const state = getRandomHexString();
+      const challenge = await generateCodeChallenge(verifier);
+      const url = getGlobusAuthorizeURL({
+        clientId,
+        scope: authorization_parameters.required_scopes,
```

Comment:
I'm concerned that there may be a security vulnerability here --- suppose the user has previously accessed:

`globus+https://good-host.com/...` which has UUID XXXX

and granted access.

Then the user gets directed to visit a Neuroglancer link that specifies a datasource of `globus+https://bad-host.com/...`   `bad-host.com` maliciously reports that it has the same UUID of XXXX as `good-host.com`.

What will happen in that case?  Will the user have to grant permission again or will it be assumed?